### PR TITLE
[Merged by Bors] - feat(Geometry/Manifold/Instances/UnitsOfNormedAlgebra): Lie group structure

### PR DESCRIPTION
**Pull request by winstonyin**
Port of [mathlib#15763](https://github.com/leanprover-community/mathlib/pull/15763)

We complete the construction of the Lie group of units of a complete normed ring.

To show that multiplication and inversion are smooth, we add some more lemmas:
1. For a manifold `M` whose chart structure is induced by an open embedding from `M` to `H`, the open embedding itself is a smooth manifold map.
2. The inverse of this open embedding is smooth on its range.
3. For a manifold map `f : M → M'` whose target space `M'` has a chart structure induced by an open embedding `e' : M' → H'`, if `e' ∘ f` is smooth, then `f` is smooth.

Lemmas previously not ported are now ported:
1. `open_embedding.to_local_homeomorph_left_inv`
2. `open_embedding.to_local_homeomorph_right_inv`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Geometry/Manifold/Instances/UnitsOfNormedAlgebra.lean
Comment (grunweg) [inline Mathlib/Geometry/Manifold/Instances/UnitsOfNormedAlgebra.lean:1]: I presume the "None of this should be particularly difficult." is resolved now; can you update this?

File: Mathlib/Geometry/Manifold/ContMDiff.lean
Comment (grunweg) [inline Mathlib/Geometry/Manifold/ContMDiff.lean:2033]: Can you pull pull this into a section? Then, you can add a variables line containing e.g.
`{M : Type*} [TopologicalSpace M] [Nonempty M] {e : M → H} (h : OpenEmbedding e) {n : WithTop ℕ}`,
prettifying this even more :-)

File: Mathlib/Geometry/Manifold/ContMDiff.lean
Comment (grunweg) [inline Mathlib/Geometry/Manifold/ContMDiff.lean:2039]: I have the nagging feeling there should be a nicer way to write this than ~a dozen underscores (say, using Lean4's named arguments). I fail to find one, though. Perhaps a reviewer has a better idea --- or you already found the best way :shrug: .

File: Mathlib/Geometry/Manifold/ContMDiff.lean
Comment (grunweg) [inline Mathlib/Geometry/Manifold/ContMDiff.lean:2050]: Here (and at a few more places in the proof) you have another opportunity to use dot notation, if you want :-)

File: Mathlib/Geometry/Manifold/ContMDiff.lean
Comment (grunweg) [inline Mathlib/Geometry/Manifold/ContMDiff.lean:2101]: As I've elaborated below, I'd use ":= by" instead.

**Review by grunweg — COMMENTED** [review]
First of all: thank you for doing this! I also work on manifolds, I'm excited to see other people doing the same. And I'm excited for mathlib to gain more instances of manifolds. This is great.

I have a couple of style suggestions: these are all small. Most of them, I presume, relate to the transition to mathlib4. (I never worked with mathlib3, so that's just a guess.) Some are my personal taste. (In any case, I'm also relatively new to mathlib, so my taste for style is still developing. Feel free to disagree with what I say.)

I've read your files out of order - sorry if that makes things confusing.

File: Mathlib/Geometry/Manifold/ContMDiff.lean
Comment (winstonyin) [reply Mathlib/Geometry/Manifold/ContMDiff.lean]: This probably requires some changes in the `ContMDiff` API, so we probably shouldn't touch it in this PR. I suppose one could give names to the type class argument `[CS : ChartedSpace H M]` so Lean4's named argument works.

File: Mathlib/Geometry/Manifold/ContMDiff.lean
Comment (grunweg) [inline Mathlib/Geometry/Manifold/ContMDiff.lean:2049]: Same here.

File: Mathlib/Geometry/Manifold/ContMDiff.lean
Comment (grunweg) [inline Mathlib/Geometry/Manifold/ContMDiff.lean:2058]: Here also, I think?

File: Mathlib/Geometry/Manifold/ContMDiff.lean
Comment (grunweg) [inline Mathlib/Geometry/Manifold/ContMDiff.lean:2077]: Here also.

File: Mathlib/Geometry/Manifold/ContMDiff.lean
Comment (grunweg) [inline Mathlib/Geometry/Manifold/ContMDiff.lean:2089]: `open`ing Set might also be useful here (I count 16 overall occurences in this file).

**Review by grunweg — COMMENTED** [review]
Thanks for the changes; I really like where this is going. I took another look, find below some more comments: small things again.
- I think the application file is almost ready to go; I suspect one lemma can just be inlined.
- The @contMDiff stuff in ContMDiff is annoying; there ought to be a better way, but I don't see it.
- Some proofs on ContMDiff are a real tour de force. Thanks for sticking with this! I feel there should be a less painful way - but perhaps that just doesn't exist yet.

Except for these (which a more experienced reviewer should comment on), from my point of view this PR is ready to go!

File: Mathlib/Geometry/Manifold/Instances/UnitsOfNormedAlgebra.lean
Comment (winstonyin) [reply Mathlib/Geometry/Manifold/Instances/UnitsOfNormedAlgebra.lean:64]: That makes sense. Inlined!

File: Mathlib/Geometry/Manifold/ContMDiff.lean
Comment (winstonyin) [reply Mathlib/Geometry/Manifold/ContMDiff.lean:2089]: Turns out it's already opened. I've removed all 16 `Set.`

File: Mathlib/Geometry/Manifold/Instances/UnitsOfNormedAlgebra.lean
Comment (grunweg) [reply Mathlib/Geometry/Manifold/Instances/UnitsOfNormedAlgebra.lean]: I [asked](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/refine.20vs.20refine'/near/403550755) and see a mild preference for `refine`, hence just pushed a commit replacing it. Hope you don't mind; feel free to revert otherwise!

**Review by grunweg — APPROVED** [review]
@maintainers From my point of view, this is good to go. Two questions of the form "is there a better way to write this" remain, namely:
- ~~Can one avoid the "_ _ _ _" dance with passing the `ChartedSpace` instance to ContMDiff? Several places. We feel one should be able to use named arguments; I tried locally and it didn't just work.~~ fixed now
- Some proofs on ContMDiff feel a bit painful to me. If there's a nicer way, I'd love to know :-) regardless of whether that's required for this PR. Update: feels better now; I'm always happy to learn new tricks ;-)

**Comment by winstonyin** [comment]
I did some testing. Seems like the `_ _ _` can be avoided simply by naming the type class arguments `[CS : ChartedSpace H M]` and then doing `(CS := …)` when calling the theorems. I’ll make an independent PR with these changes, not blocking this PR for the moment.

**Comment by grunweg** [comment]
> I did some testing. Seems like the `_ _ _` can be avoided simply by naming the type class arguments `[CS : ChartedSpace H M]` and then doing `(CS := …)` when calling the theorems.

Nice! I like the new changes; looks much cleaner. I just re-viewed everything and still think this is good to go.

**Comment by winstonyin** [comment]
Pushed one more commit. Now `ContMDiff.lean` and `SmoothManifoldWithCorners.lean` are both free of `@foo _ _ _`. Once CI goes through, we should be good to go.

**Comment by winstonyin** [comment]
@grunweg Thanks for splitting `ConMDiff.lean`. Do you think `Geometry.Manifold.ContMDiff.Atlas` is the right place for the lemmas `contMDiff_openEmbedding` etc.?

**Comment by grunweg** [comment]
You're welcome. I now know much better what mathlib already has on these :-)

When I was splitting the file, I always thought about these lemma going into `ContMDiff/Basic.lean`. I think this fits slightly better than Atlas.lean, but I don't have a very strong preference.
In any case, can you augment the module docstring to mention these new results, please?

File: Mathlib/Geometry/Manifold/ContMDiff/Basic.lean
Comment (grunweg) [inline Mathlib/Geometry/Manifold/ContMDiff/Basic.lean:29]: These are not needed, right? In this case, I'd remove them :-)
Sorry that the split made these changes a small pain to rebase.

File: Mathlib/Geometry/Manifold/ContMDiff/Defs.lean
Comment (grunweg) [inline Mathlib/Geometry/Manifold/ContMDiff/Defs.lean:60]: Same here, these can be removed, right?

File: Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean
Comment (grunweg) [inline Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean:1242]: We don't need this either, right?

File: Mathlib/Geometry/Manifold/ContMDiff/Basic.lean
Comment (winstonyin) [reply Mathlib/Geometry/Manifold/ContMDiff/Basic.lean:29]: You mean the `CS : `, etc.? I think I still need them for `ContMDiffOn.comp_contMDiff` in `ContMDiff.of_comp_openEmbedding`. Let me know if that's not the case.

File: Mathlib/Geometry/Manifold/ContMDiff/Defs.lean
Comment (winstonyin) [reply Mathlib/Geometry/Manifold/ContMDiff/Defs.lean:60]: Needed for `ContMDiffOn` and `contMDiffOn_iff` in `contMDiff_openEmbedding`.

File: Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean
Comment (winstonyin) [reply Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean:1242]: Needed in `OpenEmbedding.singleton_smoothManifoldWithCorners`

File: Mathlib/Geometry/Manifold/ContMDiff/Basic.lean
Comment (winstonyin) [reply Mathlib/Geometry/Manifold/ContMDiff/Basic.lean:31]: I suppose not all three of `CS`, `CS'`, `CS''` are useful in this PR, but I still think we should keep all three if even one of them is used somewhere.

**Comment by winstonyin** [comment]
I split out #8776 for removing `Set.`, etc., and #8780 to clean up variables in `SmoothManifoldWithCorners.lean`. Probably will keep `[CS : ChartedSpace]`, etc., in this PR.

May I say that I'm so impressed by your review, @grunweg. Every time, they improve the PR by a lot and even inspire further work.  🐙

**Review by grunweg — COMMENTED** [review]
I took another close look: in light of #8702, I presume the named `ChartedSpace` arguments ought to be simplified. I tried doing so myself, but in every lemma, something breaks in the proof. I don't fully understand why this changes and how to fix it.

I pushed two commits; feel free to revert either or both:
- the frist is golfing a proof. (This means another tiny sorry pops up, which I'd consider missing API for local homeomorphisms or so.)
- the second removes the CS arguments from two lemmas: one lemma works out nicely (and can even be golfed accordingly); for the other, a proof breaks.

File: Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean
Comment (winstonyin) [reply Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean:900]: I'm surprised Lean doesn't complain about this!

**Comment by winstonyin** [comment]
Let's try a bit to simplify the `CS` arguments in your direction. If it doesn't quite work out, we can ask on Zulip or just revert back to `_ _ _`. If anything, we may gain some nice `LocalHomeomorph` lemmas. Thanks for looking into this.

**Comment by grunweg** [comment]
I've fixed the first sorry, I'm still puzzled by the proofs breaking. Feel free to ask or revert these commits.

**Comment by grunweg** [comment]
I've pushed a commit removing all CS annotations: there's a second proof breaking, just so you see what would be needed.
Feel free to revert these if dealing with these is too painful.

**Comment by winstonyin** [comment]
I've removed the remaining `CS` annotations. How to convince Lean to use the instances we want looks to be beyond my pay grade. I'd like the PR to go through sooner, and if people are sufficiently bothered by the `_ _ _ _`, we can open a new PR to fix those. Thanks for experimenting!

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
